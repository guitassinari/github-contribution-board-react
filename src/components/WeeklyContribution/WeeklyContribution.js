import React from 'react'
import PropTypes from 'prop-types'
import DailyContribution from '../DailyContribution/DailyContribution.js'
import styled from 'styled-components'

const ContributionsContainer = styled.div`
  display: grid;
  grid-gap: 3px;
  align-content: start;
`

function WeeklyContribution({ contributions }) {
  const filteredContributions = (contributions || []).slice(0, 7) 
  return (
    <ContributionsContainer>
      {filteredContributions.map((contributionCount, index) => <DailyContribution key={index} count={contributionCount} />)}
    </ContributionsContainer>
  )
}

WeeklyContribution.propTypes = {
  contributions: PropTypes.arrayOf(PropTypes.number)
}

WeeklyContribution.defaultProps = {
  contributions: [0, 0, 0, 0, 0, 0, 0]
}

export default WeeklyContribution