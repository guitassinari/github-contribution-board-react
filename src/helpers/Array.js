export function splitArrayIntoChunksOf(chunkSize, array) {
  let index = 0;
  let arrayLength = array.length;
  let chunks = [];
  
  for (index = 0; index < arrayLength; index += chunkSize) {
      const chunk = array.slice(index, index+chunkSize);
      chunks.push(chunk);
  }

  return chunks;
}


export function unique(value, index, self) { 
  return self.indexOf(value) === index;
}