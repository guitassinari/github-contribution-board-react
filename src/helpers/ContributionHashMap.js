export default class ContributionHashMap {
  constructor(contributions) {
    this.hashMap = contributions.reduce((map, contribution) => {
      map[contribution.date] = contribution.count;
      return map;
    }, {})
  }

  getCountFor(date) {
    const key = this.__hashMapKeyFromDate(date)
    return this.hashMap[key] || 0
  }

  __hashMapKeyFromDate(date) {
    const offset = date.getTimezoneOffset()
    const utcDate = new Date(date.getTime() + (offset*60*1000))
    return utcDate.toISOString().split('T')[0]
  }
}