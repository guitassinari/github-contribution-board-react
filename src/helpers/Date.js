
export function* allDaysBetween(initialDate, finalDate) {
  let day = new Date(initialDate);
  while(day <= finalDate){
    yield day
    let nextDay = day.setDate(day.getDate() + 1);
    day = new Date(nextDay);
  }
}

export function previousMondayFrom(date) {
  let dateCopy = new Date(date);
  const day = dateCopy.getDay()
  if(day === 1) { // Already monday
    return dateCopy
  }
  
  const dayOffsetToMonday = day == 0 ? -6 : 1
  const diff = dateCopy.getDate() - day + dayOffsetToMonday;
  return new Date(dateCopy.setDate(diff));
}

export function firstDayOfYear(year) {
  return new Date(year, 0, 1);
}

export function lastDayOfTheYear(year) {
  return new Date(year, 11, 31)
}