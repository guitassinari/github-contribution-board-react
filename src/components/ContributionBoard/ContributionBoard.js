import React from 'react'
import PropTypes from 'prop-types'
import WeeklyContribution from '../WeeklyContribution/WeeklyContribution.js'
import styled from 'styled-components'
import { splitArrayIntoChunksOf } from '../../helpers/Array'

const WEEKS_PER_YEAR = 53
const NUMBER_OF_WEEK_DAYS = 7

const ContributionsContainer = styled.div`
  display: grid;
  grid-template-columns: repeat(${WEEKS_PER_YEAR}, 1fr);
  grid-gap: 3px;
  padding: 10px;
  border: 1px solid #e1e4e8;
  border-radius: 5px;
`

function ContributionBoard({ contributions }) {
  const filteredContributions = (contributions || [])
  const contributionsWeekChunks = splitArrayIntoChunksOf(NUMBER_OF_WEEK_DAYS, filteredContributions) 
  return (
    <ContributionsContainer>
      {contributionsWeekChunks.map((weekContributions, index) => <WeeklyContribution key={index} contributions={weekContributions} />)}
    </ContributionsContainer>
  )
}

ContributionBoard.propTypes = {
  contributions: PropTypes.arrayOf(PropTypes.number)
}

ContributionBoard.defaultProps = {
  contributions: []
}

export default ContributionBoard