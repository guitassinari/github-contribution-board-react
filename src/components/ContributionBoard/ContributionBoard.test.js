import React from 'react';
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme';
import ContributionBoard from './ContributionBoard';
import WeeklyContribution from '../WeeklyContribution/WeeklyContribution.js'
import { splitArrayIntoChunksOf } from '../../helpers/array'

describe('WeeklyContribution', () => {
  test('Matches snapshot', () => {
    const component = renderer.create(<WeeklyContribution />);
    const tree = component.toJSON()
  
    expect(tree).toMatchSnapshot();
  });

  describe('when it receives null contributions', () => {
    const contributions = null 

    test('Matches snapshot', () => {
      const component = renderer.create(<ContributionBoard contributions={contributions} />);
      const tree = component.toJSON()
    
      expect(tree).toMatchSnapshot();
    });

    test('renders no WeeklyContributions', () => {
      const wrapper = shallow(<ContributionBoard contributions={contributions} />);

      const renderedWeeklyContributions = wrapper.find(WeeklyContribution)
        
      expect(renderedWeeklyContributions.length).toEqual(0);
    })
  })

  describe('when receive empty contributions', () => {
    const contributions = [] 

    test('Matches snapshot', () => {
      const component = renderer.create(<ContributionBoard contributions={contributions} />);
      const tree = component.toJSON()
    
      expect(tree).toMatchSnapshot();
    });

    test('renders no WeeklyContributions', () => {
      const wrapper = shallow(<ContributionBoard contributions={contributions} />);

      const renderedWeeklyContributions = wrapper.find(WeeklyContribution)
        
      expect(renderedWeeklyContributions.length).toEqual(0);
    })
  })

  describe('when it receives contributions', () => {
    const contributions = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
    const weeklyContributions = splitArrayIntoChunksOf(7, contributions)

    test('Matches snapshot', () => {
      const component = renderer.create(<ContributionBoard contributions={contributions} />);
      const tree = component.toJSON()
    
      expect(tree).toMatchSnapshot();
    });

    test('renders weeks of contributions', () => {
      const wrapper = shallow(<ContributionBoard contributions={contributions} />);

      const renderedWeeklyContributions = wrapper.find(WeeklyContribution)
        
      expect(renderedWeeklyContributions.length).toEqual(weeklyContributions.length);

      renderedWeeklyContributions.forEach((weeklyContribution, index) => {
        expect(weeklyContribution.props().contributions).toEqual(weeklyContributions[index])
      })
    })
  })
})
