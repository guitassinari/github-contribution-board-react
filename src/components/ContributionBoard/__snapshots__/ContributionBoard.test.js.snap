// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeeklyContribution Matches snapshot 1`] = `
<div
  className="sc-gtssRu rRoJ"
>
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
  <div
    className="sc-bdnylx gPrTLT"
  />
</div>
`;

exports[`WeeklyContribution when it receives contributions Matches snapshot 1`] = `
<div
  className="sc-dlnjPT bTrRUP"
>
  <div
    className="sc-gtssRu rRoJ"
  >
    <div
      className="sc-bdnylx bkFITR"
    />
    <div
      className="sc-bdnylx bkFITR"
    />
    <div
      className="sc-bdnylx fFbPkp"
    />
    <div
      className="sc-bdnylx fFbPkp"
    />
    <div
      className="sc-bdnylx fFbPkp"
    />
    <div
      className="sc-bdnylx eankiZ"
    />
    <div
      className="sc-bdnylx eankiZ"
    />
  </div>
  <div
    className="sc-gtssRu rRoJ"
  >
    <div
      className="sc-bdnylx eankiZ"
    />
    <div
      className="sc-bdnylx dSDfEd"
    />
    <div
      className="sc-bdnylx dSDfEd"
    />
    <div
      className="sc-bdnylx dSDfEd"
    />
    <div
      className="sc-bdnylx dSDfEd"
    />
    <div
      className="sc-bdnylx dSDfEd"
    />
    <div
      className="sc-bdnylx dSDfEd"
    />
  </div>
  <div
    className="sc-gtssRu rRoJ"
  >
    <div
      className="sc-bdnylx dSDfEd"
    />
  </div>
</div>
`;

exports[`WeeklyContribution when it receives null contributions Matches snapshot 1`] = `
<div
  className="sc-dlnjPT bTrRUP"
/>
`;

exports[`WeeklyContribution when receive empty contributions Matches snapshot 1`] = `
<div
  className="sc-dlnjPT bTrRUP"
/>
`;
